src/creation/creator.rs

```diff
@@ -0,0 +1,268 @@
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Creator as GenericCreator;
+    use crate::{
+        testing::mock::{Data, Hasher64},
+        units::{
+            FullUnit as GenericFullUnit, PreUnit as GenericPreUnit, Unit as GenericUnit, UnitCoord,
+        },
+        Hasher, NodeCount, NodeIndex, Round,
+    };
+    use std::collections::HashSet;
+
+    type Creator = GenericCreator<Hasher64>;
+    type PreUnit = GenericPreUnit<Hasher64>;
+    type Unit = GenericUnit<Hasher64>;
+    type FullUnit = GenericFullUnit<Hasher64, Data>;
+
+    fn creator_set(n_members: NodeCount) -> Vec<Creator> {
+        let mut result = Vec::new();
+        for i in 0..n_members.0 {
+            result.push(Creator::new(NodeIndex(i), n_members));
+        }
+        result
+    }
+
+    fn create_units<'a, C: Iterator<Item = &'a Creator>>(
+        creators: C,
+        round: Round,
+    ) -> Vec<(PreUnit, Vec<<Hasher64 as Hasher>::Hash>)> {
+        let mut result = Vec::new();
+        for creator in creators {
+            result.push(
+                creator
+                    .create_unit(round)
+                    .expect("Creation should succeed."),
+            );
+        }
+        result
+    }
+
+    fn preunit_to_unit(preunit: PreUnit) -> Unit {
+        FullUnit::new(
+            preunit,
+            // The coord is wrong, but it doesn't matter.
+            Data::new(UnitCoord::new(0, 0.into()), 0),
+            0,
+        )
+        .unit()
+    }
+
+    fn add_units(creator: &mut Creator, units: &[Unit]) {
+        for unit in units {
+            creator.add_unit(unit);
+        }
+    }
+
+    #[test]
+    fn creates_initial_unit() {
+        let n_members = NodeCount(7);
+        let round = 0;
+        let creator = Creator::new(NodeIndex(0), n_members);
+        assert_eq!(creator.current_round(), round);
+        let (preunit, parent_hashes) = creator
+            .create_unit(round)
+            .expect("Creation should succeed.");
+        assert_eq!(preunit.round(), round);
+        assert_eq!(parent_hashes.len(), 0);
+    }
+
+    #[test]
+    fn creates_unit_with_all_parents() {
+        let n_members = NodeCount(7);
+        let mut creators = creator_set(n_members);
+        let new_units = create_units(creators.iter(), 0);
+        let new_units: Vec<_> = new_units
+            .into_iter()
+            .map(|(pu, _)| preunit_to_unit(pu))
+            .collect();
+        let expected_hashes: Vec<_> = new_units.iter().map(|u| u.hash()).collect();
+        let creator = &mut creators[0];
+        add_units(creator, &new_units);
+        let round = 1;
+        assert_eq!(creator.current_round(), 0);
+        let (preunit, parent_hashes) = creator
+            .create_unit(round)
+            .expect("Creation should succeed.");
+        assert_eq!(preunit.round(), round);
+        assert_eq!(parent_hashes, expected_hashes);
+    }
+
+    #[test]
+    fn creates_unit_with_minimal_parents() {
```

Comment:
I think this test is very useful. Could you make it run not only for n=7 p=5, but also for (n,p) in {(1, 1), (2, 2), (3, 3), (4, 3), (5, 4), (6, 5)}?
The same for `cannot_create_unit_below_parents_threshold` :)

Reply:
Hah ok :)